### PR TITLE
Add depth limiting and path globs options

### PR DESCRIPTION
**Pull request by calebeby**
# Setup

`git checkout depth-limiting-and-matchers (&& npm install) && npm run build && npm link`

## `--max-depth <number>`

Limits the depth of crawling. 1 means only fetch the page you pass in, 2 means fetch the page you pass in and any page it directly links to, etc.

Example: `lighthouse-parade https://www.thebluealliance.com/team/2733 --max-depth 2`

## `--path-must-match <string>`

Allows specifying a glob. Any URLs that do not match this glob will not be fetched.

Example: `lighthouse-parade "https://www.thebluealliance.com/team/2733/2010" --max-depth 2 --path-must-match "/team/*/2010"`

You can also pass multiple globs to allow by passing the flag multiple times: `lighthouse-parade "https://www.thebluealliance.com/team/2733/2010" --max-depth 2 --path-must-match "/team/*/2010" --path-must-match "/event/*"`

## `--path-must-not-match <string>`

Allows specifying a glob. Any URLs that match this glob will not be fetched (higher precedence than `--path-must-match`)

Example: `lighthouse-parade "https://en.wikipedia.org" --path-must-not-match "/wiki/File:*"`

### CODE REVIEW

**Comment by emersonthis** [comment]
This may not be new to this PR, but I notice that when the scan doesn't find anything, the script just hangs. It says "scan complete" but you never get the prompt back and have to force-quit the script. If it's not too hard to do, I think we should fix that in this PR.

**Comment by emersonthis** [comment]
I'm seeing some unexpected results with the `--path-must-not-match`...
Screen Shot 2020-12-23 at 10 03 22 AM

Also with `--path-must-match`...
Screen Shot 2020-12-23 at 10 07 47 AM
I was expecting both of these ^ to include https://cloudfour.com/thinks/cooperative-design-systems because it's linked from the homepage. Am I misunderstanding something?

**Comment by emersonthis** [comment]
Outside the context of the (maybe) bugs above, I have some initial usability thoughts:
* For me it took several tries before I realized I wasn't quoting the glob. Maybe that's not essential for some values, but we should work to make that really clear. Or maybe it's just me.
* I also suspect users might try to include the domain, even though the flag has "path" in the name. Ex: `--path-must-match "https://cloudfour.com/thinks/*`. (I did this a few times.) Would it make sense to add logic that handles this?
* Regarding the names of the flags, I wonder if we name them a bit more defensively/specifically. Ex: `--max-crawl-depth`? `--path-must-match-glob`? I thinking this might be a bit more future-proof (in case we want to add new options that do similar things) and also maybe a little more user-friendly.

File: README.md [hunk 1]
```diff
@@ -33,11 +33,14 @@ Runs a crawler on the provided URL. Discovers all URLs and runs a lighthouse rep
...
---ignore-robots             Crawl pages even if they are listed in the site's robots.txt  (default: false)
+--ignore-robots             Crawl pages even if they are listed in the site's robots.txt  (default false)
 --crawler-user-agent        Pass a user agent string to be used by the crawler (not by Lighthouse)
 --lighthouse-concurrency    Control the maximum number of ligthhouse reports to run concurrently  (default: number of CPU cores minus one)
+--max-depth                 Control the maximum depth of crawled links
+--path-must-match           Specify a glob (in quotes) for paths to match. Links to non-matched paths will not be crawled. This flag can be specified multiple times to allow multiple paths
+--path-must-not-match       Specify a glob (in quotes) for paths to exclude. Links to matched paths will not be crawled. This flag can be specified multiple times to exclude multiple paths
 -v, --version               Displays current version
--h, --help                  Displays help text
+-h, --help                  Displays this message
```
Comment (emersonthis) [inline README.md:43]: Is it possible this message appears in different contexts, besides the CLI? Might we use it in our documentation, etc? I'm not too worried about people knowing what the `--help` option does...

File: cli.ts [hunk 2]
```diff
@@ -25,8 +25,14 @@ const symbols = {
   success: kleur.green('✔'),
 };
 
+const toArray = <T extends unknown>(input: T) =>
+  Array.isArray(input) ? input : [input];
+
 sade('lighthouse-parade <url> [dataDirectory]', true)
   .version(version)
+  .example(
+    '"https://www.thebluealliance.com/team/2733/2010" --max-depth 2 --path-must-match "/team/*/2010" --path-must-match "/event/*"'
```
Comment (emersonthis) [inline cli.ts:34]: I think robots are cool, but I propose we converge on a consistent pattern for URL examples. Using https://cloudfour.com is my first instinct, but I'm open to other ideas.

File: cli.ts [hunk 3]
```diff
@@ -44,6 +50,15 @@ sade('lighthouse-parade <url> [dataDirectory]', true)
     'Control the maximum number of ligthhouse reports to run concurrently',
     os.cpus().length - 1
   )
+  .option('--max-depth', 'Control the maximum depth of crawled links')
+  .option(
+    '--path-must-match',
+    'Specify a glob (in quotes) for paths to match. Links to non-matched paths will not be crawled. This flag can be specified multiple times to allow multiple paths'
```
Comment (emersonthis) [inline cli.ts:56]: I wonder if we should add language to clarify that the initial entry point URL is an exception. Correct? If I do `lighthouse-parade https://cloudfour.com --path-must-match "/foofoodoesntexist"` I will _still_ receive one report for the homepage, right? I can't tell if that would be obvious or helpful.

File: test/path-filter-globs.test.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+});
+
+test('allow items from multiple include globs', () => {
+  const filter = createUrlFilter(['/foo/*', '/foo'], []);
+  expect(filter({ path: '/foo' })).toBeTruthy();
+  expect(filter({ path: '/foo/bar' })).toBeTruthy();
+  expect(filter({ path: '/foo/bar/baz' })).toBeFalsy();
+  expect(filter({ path: '/asdf' })).toBeFalsy();
+});
+
```
Comment (emersonthis) [inline test/path-filter-globs.test.ts:32]: Is `/foo/*/bar` currently supported?

File: test/path-filter-globs.test.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+test('exclude has higher precedence than include', () => {
```
Comment (emersonthis) [inline test/path-filter-globs.test.ts:33]: Smart test!! Nice work

File: test/path-filter-globs.test.ts (hunk 4 above)
Comment (calebeby) [reply test/path-filter-globs.test.ts:32]: Yes. I added a test.

File: README.md (hunk 1 above)
Comment (calebeby) [reply README.md:43]: Oops I just copy pasted from what gets displayed when I do `--help`. Do you want me to remove that from the README?

**Comment by calebeby** [comment]
@emersonthis How do you suggest making it more clear that quotes are needed for the globs? The help text for `--path-must-match` already says "Specify a glob (in quotes) ..."

File: README.md (hunk 1 above)
Comment (emersonthis) [reply README.md:43]: I think it's fine to include that in the README. Probably an efficient way to keep it up-to-date.

**Comment by calebeby** [comment]
@emersonthis I don't have a strong opinion on the max-crawler depth flag. I think that the way it is makes sense. I think of it as: depth is 1, so there is 1 "tier" of pages.

**Comment by calebeby** [comment]
@emersonthis Can you take another look? I haven't updated the help text in the README yet but I will before I merge

**Review by emersonthis — APPROVED** [review]

**Pull request merged**